Fix load of models that depend on non thread-safe dependencies

paulojrp opened the pull request:
A problem was detected when loading TensorFlow models in different threads inside the same JVM. That happened after load a TensorFlow model and then try to import a new TensorFlow model. This was caused by a dependency of TensorFlow (protobuf) that was being reloaded but it already existed in the JVM (through the 1st thread).

The workaround was to share the problematic module (Tensorflow) across the sub-interpreters of Python. This is a workaround for the issues with CPython extensions.

paulojrp commented:
at the moment I only tested this with UTs, before landing this I want to make sure that it works in a real environment

pedrorijo91 commented:
I suppose this is related with #26 ?

pedrorijo91 commented on openml-python-common/src/main/java/com/feedzai/openml/python/jep/instance/JepInstance.java:
```diff
@@ -62,9 +62,56 @@
+                        // Start the Jep instance thread
+                        jepInstance.start();
+                    }catch(Exception e){
+                        throw new RuntimeException("Exception occurred in creating singleton jep instance");
```
add caught exception to the created RuntimeException

pedrorijo91 commented on openml-python-common/src/main/java/com/feedzai/openml/python/jep/instance/JepInstance.java:
```diff
@@ -62,9 +62,56 @@
+    /**
+     * Private constructor for singleton pattern.
+     *
+     * @implNote {@link Jep} is initalized inside the current JVM. Due to the need to manage a consistent Python thread
```
kudos for the explanation! IMHO I would prefer to keep the explanation generic instead of base it on the specific case we caught (and then with a reference to the specific github issue), but it's a super minor nitpick, and please let anyone else comment on this before changing anything

pedrorijo91 reviewed (commented):
A quick review seems good. will make a deeper review monday. But there are tests failing

nmldiegues commented:
A clear disadvantage of this approach is that it never "releases" models already loaded. E.g., consider a JVM where you load and close 10000 models sequentially; this will eventually go OOM.
I'm thinking: could you maybe cache the model loading in the ClassificationPythonModel, and in the "close" just "null out" the variable (for Python GC to kick in)? That would probably make this solution perfect.

Quick note: the docker image used in travis is not set up with tensorflow, but I see you added as UT the regression test for it. Nevertheless there are some failed UTs. I'll let you take a look at it.

paulojrp commented:
> I suppose this is related with #26 ?

yes this commit fixed that  issue

paulojrp commented:
> A clear disadvantage of this approach is that it never "releases" models already loaded. E.g., consider a JVM where you load and close 10000 models sequentially; this will eventually go OOM.
> I'm thinking: could you maybe cache the model loading in the ClassificationPythonModel, and in the "close" just "null out" the variable (for Python GC to kick in)? That would probably make this solution perfect.
> 
> Quick note: the docker image used in travis is not set up with tensorflow, but I see you added as UT the regression test for it. Nevertheless there are some failed UTs. I'll let you take a look at it.

Another problem is the names of to methods responsible for the classification of events. The implementation of the providers is assuming that their names are constantes ("classify" and "getClassification"). This logic won't work if we only use one python interpreter. I will explore another solutions.

Commit:
Fix load of models that depend on non thread-safe dependencies

A problem was detected when loading TensorFlow models in different threads inside the same JVM. That happened after load a TensorFlow model and then try to import a new TensorFlow model. This was caused by a dependency of TensorFlow (protobuf) that was being reloaded but it already existed in the JVM (through the 1st thread).

The workaround was to share the problematic module (Tensorflow) across the sub-interpreters of Python. This is a workaround for the issues with CPython extensions.

nmldiegues commented:
This is pretty nice @paulojrp . Of course it has the downside that it requires code changes for every new library with the same problem, but it is an acceptable trade-off. I trust that you ran the UT regression of https://github.com/feedzai/feedzai-openml-python/issues/26 and it passed?

I just have 1 suggestion:  Please make the list of packages added to shared modules read dynamically (environment variable? a resource file inside the openml jar? please consider remote spark executions when devising the solution) --- this will make users autonomous to overcome this problem for a new library in the future; please add a note to the README about the https://github.com/feedzai/feedzai-openml-python/issues/26 and this solution

pedrorijo91 commented on openml-python-common/README.md:
````diff
@@ -40,3 +40,15 @@ export PATH=$ANACONDA_PATH/envs/myenv/bin:$PATH
 export LD_PRELOAD=$ANACONDA_PATH/envs/myenv/lib/libpython3.6m.so
 ```
+
+7. If you need to share Python modules across sub-interpreters, you would need to create a "python-packages.xml" file where you define the modules to be shared. By default the provider is already sharing the "numpy" and "tensorflow" modules. This is a workaround for the issues with CPython extensions.
````
I think it should be 'you will need to create a (...)'. @krisztinaknagy to validate this one

pedrorijo91 commented on openml-python-common/README.md:
````diff
@@ -40,3 +40,15 @@ export PATH=$ANACONDA_PATH/envs/myenv/bin:$PATH
 export LD_PRELOAD=$ANACONDA_PATH/envs/myenv/lib/libpython3.6m.so
 ```
+
+7. If you need to share Python modules across sub-interpreters, you would need to create a "python-packages.xml" file where you define the modules to be shared. By default the provider is already sharing the "numpy" and "tensorflow" modules. This is a workaround for the issues with CPython extensions.
````
is there any official documentation for this python-packages.xml?

paulojrp replied:
bellow you can find an example of the contents of this file

pedrorijo91 commented on openml-python-common/src/main/java/com/feedzai/openml/python/jep/instance/JepInstance.java:
```diff
@@ -97,8 +101,18 @@ public void stop() {
+
+        final JepConfig jepConfig = new JepConfig()
+                .addSharedModules(sharedModules.toArray(new String[0]))
+                .setInteractive(false);
```
what is this interactive?

pedrorijo91 replied:
ok, got it: https://github.com/ninia/jep/blob/master/src/main/java/jep/JepConfig.java#L65

is this relevant for this to work or just a minor optimization? Asking because it seems it shouldn't be needed, and we didn't have it before.

paulojrp replied:
I'm not sure about the implication of changing this behaviour. I'm using the same behaviour used before this commit.

pedrorijo91 replied:
oh, right, the interactive flag was passed directly to Jep constructor. you are right, forget my comment :)

pedrorijo91 commented on openml-python-common/src/main/java/com/feedzai/openml/python/xml/parser/XMLParser.java:
```diff
@@ -0,0 +1,104 @@
+ * @author Paulo Pereira (paulo.pereira@feedzai.com)
+ * @since 0.1.5
+ */
+public class XMLParser {
```
I would rename for SharedModulesParser or something like that, since the main goal of this class is to retrieve the shared modules no? XML is just a impl detail right?

Or do you designed this class with the possibility of parsing other XML files?

pedrorijo91 commented on openml-python-common/src/main/java/com/feedzai/openml/python/xml/parser/XMLParser.java:
```diff
@@ -0,0 +1,104 @@
+     * Constructor.
+     */
+    public XMLParser() {
+        this.xmlFile = "python-packages.xml";
```
not sure if we would like to allow other filenames? If yes, then we could have an overloaded constructor for that. If not, I think the filename could be a class constant?

pedrorijo91 commented on openml-python-common/src/main/java/com/feedzai/openml/python/xml/parser/XMLParser.java:
```diff
@@ -0,0 +1,104 @@
+     * @return A {@link Set} with the modules of Python to be shared across sub-interpreters.
+     * @throws Exception If there is an error while parsing {@link #xmlFile}.
+     */
+    private Set<String> parseXMLFile() throws Exception {
```
can we be more specific on which Exception(s) are thrown?

paulojrp replied:
yes we could but this method throws three different exceptions and there is nothing to gain from specifying each of those exceptions

pedrorijo91 replied:
some would disagree :p https://pmd.github.io/pmd-6.0.0/pmd_rules_java_design.html#signaturedeclarethrowsexception

but I would really vouch if we had 1 or 2 exceptions. 3 is a threshold where I feel ok in declaring simply Exception :)

pedrorijo91 commented on openml-python-common/src/main/java/com/feedzai/openml/python/xml/parser/XMLParser.java:
```diff
@@ -0,0 +1,104 @@
+     * @throws Exception If there is an error while parsing {@link #xmlFile}.
+     */
+    private Set<String> parseXMLFile() throws Exception {
+        final HashSet<String> sharedModules = new HashSet<>();
```
Do you really need sharedModules to be an HashSet, or any Set will do? I suspect any Set would do: in that case, declare the variable as Set<String> pls

pedrorijo91 replied:
furthermore, I'm not very confortable in returning a mutable collection. What do you think of using an ImmutableSet.Builder ?

pedrorijo91 commented on openml-python-common/src/main/java/com/feedzai/openml/python/xml/parser/XMLParser.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    /**
+     * Retrieves a {@link Set} with the modules of Python to be shared across sub-interpreters.
```
"A set with the name of the Python modules" would be more correct

pedrorijo91 commented on openml-python-common/src/test/java/com/feedzai/openml/python/xml/parser/XMLParserTest.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    /**
+     * Tests the retrieving of the shared modules from an invalid XML file.
```
what about an unexisting xml file?

pedrorijo91 commented on openml-python-common/src/test/java/com/feedzai/openml/python/xml/parser/XMLParserTest.java:
```diff
@@ -0,0 +1,82 @@
+    private static final Logger logger = LoggerFactory.getLogger(XMLParserTest.class);
+
+    /**
+     * Tests the retrieving of the shared modules from a valid XML file.
```
valid means with the correct format right?

pedrorijo91 reviewed (commented):
overall looks great! :) just a few minor style issues. Feel free to disagree with them

pedrorijo91 reviewed (commented):
actually, I would expect a UT like the one described in #26 . Is it too hard to automate?

nmldiegues commented on openml-python-common/src/main/java/com/feedzai/openml/python/jep/instance/JepInstance.java:
```diff
@@ -97,8 +101,18 @@ public void stop() {
-        try (final Jep jep = new Jep(false)) {
+        final Set<String> sharedModules = ImmutableSet.<String>builder()
+                .add("tensorflow")
+                .add("numpy")
```
Please put these modules in a default XML file instead?

That way the user can remove them (e.g., what if he does not have tensorflow installed?)

paulojrp replied:
I prefer to include this modules hardcoded. The module doesn't have to be installed to be declared as a shared module (e.g. this test was executed in CI and the docker image used for that doesn't contain tensorflow installed).

Also, by having these modules hardcoded we wouldn't have to add the xml file to the library directories of the repositories that depend on this module.

nmldiegues commented on openml-python-common/src/main/java/com/feedzai/openml/python/modules/SharedModulesParser.java:
```diff
@@ -0,0 +1,112 @@
+     * Constructor.
+     */
+    public SharedModulesParser() {
+        this("python-packages.xml");
```
Constant variable for the string

nmldiegues commented on openml-python-common/src/main/java/com/feedzai/openml/python/modules/SharedModulesParser.java:
```diff
@@ -0,0 +1,112 @@
+        }
+        return sharedModules;
+    }
+}
```
Now this was already done, but it was a really bit overkill :)   I'd had used just 1 package per line, or a JSON file. But since it's done, let's go for it

paulojrp replied:
we only have one package per line

e.g.
```
<?xml version="1.0"?>
<python>
   <package>my_package_1</package>
   <package>my_package_2</package>
</python>
```

nmldiegues replied:
I meant something much simpler:
```
my_package_1
my_package_2
```

But it's alright!

nmldiegues commented on openml-python-common/src/main/java/com/feedzai/openml/python/modules/package-info.java:
```diff
@@ -0,0 +1,23 @@
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
```
Same as above

nmldiegues commented on openml-python-common/src/main/java/com/feedzai/openml/python/modules/SharedModulesParser.java:
```diff
@@ -0,0 +1,112 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
```
Please make sure this copyright is the same as in other files in this repo

paulojrp replied:
?
it's the same

nmldiegues replied:
It was a request to make sure that was the case :)

nmldiegues commented on openml-python-common/src/test/java/com/feedzai/openml/python/modules/SharedModulesParserTest.java:
```diff
@@ -0,0 +1,71 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
```
Same as above

nmldiegues reviewed (commented):
@pedrorijo91  for adding that UT we'd need tensorflow to be added to the Docker image used in Travis (and then to the base openml image used inside Feedzai). I'm not sure we want to keep adding all the stuff to the images, they'll grow endlessly.

@paulojrp Reviewed, mostly minor comments: please check the coverage and improve it, and tackle my comments. Thanks!

pedrorijo91 commented:
can't we simply create a custom model that loads python stuff @nmldiegues ? anyway, if it's too time/effort-consuming, let's be pragmatic and go without it :)

paulojrp commented:
@nmldiegues I created a new UT but the coverage is still decreasing a lot, I don't see how this is related with my modifications.

@pedrorijo91 to reproduce this problem we only had to load a python file that imports tensorflow. The problem with a unit-test for this case is that we would have to modify the docker image to also install tensorflow. we could do it but then what would happen when new packages with these problems are found? (modify the image to also check that). Don't forget that we don't have an openml provider for tensorflow.
Also this modification was tested manually by loading an tensorflow model and by using it to score events.

pedrorijo91 reviewed (approved):
all good on my side then :)
